src/Storages/MaterializedView/RefreshTask.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <Storages/MaterializedView/RefreshSet.h>
+#include <Storages/MaterializedView/RefreshTask_fwd.h>
+#include <Storages/MaterializedView/RefreshSchedule.h>
+#include <Storages/MaterializedView/RefreshSettings.h>
+
+#include <Core/BackgroundSchedulePool.h>
+
+#include <random>
+
+
+namespace DB
+{
+
+class PipelineExecutor;
+
+class StorageMaterializedView;
+class ASTRefreshStrategy;
+
+class RefreshTask : public std::enable_shared_from_this<RefreshTask>
+{
+public:
+    /// Never call it manually, public for shared_ptr construction only
+    explicit RefreshTask(const ASTRefreshStrategy & strategy);
+
+    /// The only proper way to construct task
+    static RefreshTaskHolder create(
+        const StorageMaterializedView & view,
+        ContextMutablePtr context,
+        const DB::ASTRefreshStrategy & strategy);
+
+    void initializeAndStart(std::shared_ptr<StorageMaterializedView> view);
+
+    /// Call when renaming the materialized view.
+    void rename(StorageID new_id);
+
+    /// Call when changing refresh params (ALTER MODIFY REFRESH).
+    void alterRefreshParams(const DB::ASTRefreshStrategy & new_strategy);
+
+    RefreshInfo getInfo() const;
+
+    /// Enable task scheduling
+    void start();
+
+    /// Disable task scheduling
+    void stop();
+
+    /// Schedule task immediately
+    void run();
+
+    /// Cancel task execution
+    void cancel();
+
+    /// Permanently disable task scheduling and remove this table from RefreshSet.
+    void shutdown();
+
+    /// Notify dependent task
+    void notify(const StorageID & parent_id, std::chrono::sys_seconds parent_next_prescribed_time);
+
+    /// For tests
+    void setFakeTime(std::optional<Int64> t);
+
+private:
+    Poco::Logger * log = nullptr;
+    std::weak_ptr<IStorage> view_to_refresh;
```

Comment:
@al13n321, weak pointers to IStorage objects are not allowed because we use `std::shared_ptr::unique()` with `StoragePtr` (and there's an unavoidable race condition with `std::weak_ptr::lock()`)